zuul-core/src/main/java/com/netflix/netty/common/throttle/RejectionUtils.java

```diff
@@ -0,0 +1,270 @@
+     * @param rejectionHeaders additional HTTP headers to add to the rejection response
+     */
+    public static void sendRejectionResponse(
+            ChannelHandlerContext ctx, StatusCategory nfStatus, String reason, HttpRequest request,
+            @Nullable Integer injectedLatencyMillis, HttpResponseStatus rejectedCode, String rejectedBody,
+            Map<String, String> rejectionHeaders) {
+        boolean shouldClose = closeConnectionAfterReject(ctx.channel());
+        // Write out a rejection response message.
+        FullHttpResponse response = createRejectionResponse(rejectedCode, rejectedBody, shouldClose, rejectionHeaders);
+
+        if (injectedLatencyMillis != null && injectedLatencyMillis > 0) {
+            // Delay writing the response for configured time.
+            ctx.executor().schedule(() -> {
+                CurrentPassport.fromChannel(ctx.channel()).add(PassportState.IN_REQ_REJECTED);
+                ctx.writeAndFlush(response);
+            }, injectedLatencyMillis, TimeUnit.MILLISECONDS);
+        } else {
+            // Write the response immediately.
+            CurrentPassport.fromChannel(ctx.channel()).add(PassportState.IN_REQ_REJECTED);
+            ctx.writeAndFlush(response);
+        }
+
+        // Notify other handlers that we've rejected this request.
+        notifyHandlers(ctx, nfStatus, rejectedCode, reason, request);
+    }
+
+    /**
+     * Marks the given channel for being closed after the next response.
+     *
+     * @param ctx the channel handler processing the request
+     */
+    public static void allowThenClose(ChannelHandlerContext ctx) {
+        // Just flag this channel to be closed after response complete.
+        ctx.channel().attr(ConnectionCloseChannelAttributes.CLOSE_AFTER_RESPONSE).set(ctx.newPromise());
+
+        // And allow this request through without rejecting.
+    }
+
+    /**
+     * Throttle either by sending rejection response message, or by closing the connection now, or just drop the
+     * message. Only call this if ThrottleResult.shouldThrottle() returned {@code true}.
+     *
+     * @param ctx the channel handler processing the request
+     * @param msg the request that is being rejected.
+     * @param rejectionType the type of rejection
+     * @param nfStatus the status to use for metric reporting
+     * @param reason the reason for rejecting the request.  This is not sent back to the client.
+     * @param injectedLatencyMillis optional parameter to delay sending a response. The reject notification is still
+     *                              sent up the pipeline.
+     * @param rejectedCode the HTTP code to send back to the client.
+     * @param rejectedBody the HTTP body to be sent back.  It is assumed to be of type text/plain.
+     * @param rejectionHeaders additional HTTP headers to add to the rejection response
+     */
+    public static void handleRejection(
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 167-178 into a separate method.